File: leetcode/108.convert-sorted-array-to-bst/ts/sortedarraytobst.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const sortedArrayToBST = (nums: number[]): TreeNode | null => {
    return traversal(nums, 0, nums.length)
}

const traversal = (nums: number[], left: number, right: number): TreeNode | null => {
    if (left >= right) return null

    const mid = left + Math.floor((right - left) / 2)
    const root = new TreeNode(nums[mid])

    root.left = traversal(nums, left, mid)
    root.right = traversal(nums, mid + 1, right)

    return root
}

export default sortedArrayToBST

File: leetcode/108.convert-sorted-array-to-bst/go/sortedarraytobst.go
package sortedarraytobst

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	return traversal(nums, 0, len(nums))
}

func traversal(nums []int, left int, right int) *TreeNode {
	if left >= right {
		return nil
	}
	
	mid := left + (right - left) / 2
	
	root := TreeNode{Val: nums[mid]}

	root.Left = traversal(nums, left, mid)
	root.Right = traversal(nums, mid + 1, right)

	return &root
}
